peps/pep-0733.rst

```diff
@@ -0,0 +1,670 @@
+returning ``0`` for success and ``-1`` for failure. For
+example, ``PyArg_ParseTuple`` returns 0 for success and
+non-zero for failure
+[`Issue 25 <https://github.com/capi-workgroup/problems/issues/25>`__].
+
+The macros ``Py_CLEAR`` and ``Py_SETREF`` access their arg more than
+once, so if the arg is an expression with side effects, they are
+duplicated
+[`Issue 3 <https://github.com/capi-workgroup/problems/issues/3>`__].
+
+The meaning of ``Py_SIZE`` depends on the type and is not always
+reliable
+[`Issue 10 <https://github.com/capi-workgroup/problems/issues/10>`__].
+
+Some API function do not have the same behaviour as their Python
+equivalents.  The behaviour of ``PyIter_Next`` is different from
+``tp_iternext``.
+[`Issue 29 <https://github.com/capi-workgroup/problems/issues/29>`__].
+The behaviour of ``PySet_Contains`` is different from ``set.__contains__``
+[`Issue 6 <https://github.com/capi-workgroup/problems/issues/6>`__].
+
+The fact that ``PyArg_ParseTupleAndKeywords`` takes a non-const
+``char*`` array as argument makes it more difficult to use
+[`Issue 28 <https://github.com/capi-workgroup/problems/issues/28>`__].
+
+``Python.h`` does not expose the whole API. Some headers (like ``marshal.h``)
+are not included from ``Python.h``.
+[`Issue 43 <https://github.com/capi-workgroup/problems/issues/43>`__].
+
+**Naming**
+
+``PyLong`` and ``PyUnicode`` use names which no longer match the Python
+types they represent (``int``/``str``). This can be fixed in a new API
+[`Issue 14 <https://github.com/capi-workgroup/problems/issues/14>`__].
+
+There are identifiers in the API which are lacking a ``Py``/``_Py``
+prefix
+[`Issue 46 <https://github.com/capi-workgroup/problems/issues/46>`__].
+
+Missing Functionality
+---------------------
+
+This section consists of a list of feature requests, i.e., functionality
+that was identified as missing in the current C API.
+
+Debug Mode
+~~~~~~~~~~
+
+A debug mode that can be activated without recompilation and which
+activates various checks that can help detect various types of errors
+[`Issue 36 <https://github.com/capi-workgroup/problems/issues/36>`__].
+
+Introspection
+~~~~~~~~~~~~~
+
+There aren't currently reliable introspection capabilities for objects
+defined in C in the same way as there are for Python objects
+[`Issue 32 <https://github.com/capi-workgroup/problems/issues/32>`__].
+
+Efficient type checking for heap types
+[`Issue 17 <https://github.com/capi-workgroup/problems/issues/17>`__].
+
+Improved Interaction with Other Languages
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Interfacing with other GC based languages, and integrating their
+GC with Python's GC
+[`Issue 19 <https://github.com/capi-workgroup/problems/issues/19>`__].
+
+Inject foreign stack frames to the traceback
+[`Issue 18 <https://github.com/capi-workgroup/problems/issues/18>`__].
+
+Concrete strings that can be used in other languages
+[`Issue 16 <https://github.com/capi-workgroup/problems/issues/16>`__].
+
```

Comment:
It may be useful to add a section "Future Developments" which mentions things that will in one form or another affect the future C API and are on the roadmap, e.g. the no-GIL patches, or possible future changes such as moving away from reference counting, JIT compilation, sharing Python objects between different in-process interpreter instances.

Reply:
Mentioning that no-GIL is in progress against the current API seems like helpful context to capture here.

More speculative changes will get their own discussions and PEPs. There's more than enough speculation earlier in the doc already...

Reply:
We could mention this in the intro: where we talk about the history, say that there will be more developments coming.

Reply:
Agreed and I believe that's a good idea in general to not make it look like this document will be the only one defining the future of the Python C API.